src/python/pants/bsp/goal.py

```diff
@@ -26,15 +38,140 @@ class BSPGoal(BuiltinGoal):
+            "--server",
+            type=bool,
+            default=False,
+            advanced=True,
+            help=(
+                "Run the Build Server Protocol server. Pants will receive RPC requests via the console. "
+                "This should only be invoked via the IDE."
+            ),
+        )
+        register(
+            "--runner-env-vars",
+            type=list,
+            member_type=str,
+            default=["PATH"],
+            help=(
+                "Environment variables to set in the BSP runner script when setting up BSP support. "
+                "Entries are either strings in the form `ENV_VAR=value` to set an explicit value; "
+                "or just `ENV_VAR` to copy the value from Pants' own environment.\n\n"
+                "Note: IntelliJ on macOS apparently does not pass through the PATH from the shell, "
+                "and so on macOS at the very least, `PATH` should be explicitly written into the BSP "
+                "runner script via this option."
+            ),
+            advanced=True,
+        )
+
     def run(
         self,
         *,
         build_config: BuildConfiguration,
         graph_session: GraphSession,
         options: Options,
         specs: Specs,
-        union_membership: UnionMembership
+        union_membership: UnionMembership,
+    ) -> ExitCode:
+        goal_options = options.for_scope(self.name)
+        if goal_options.server:
+            return self._run_server(
+                graph_session=graph_session,
+                union_membership=union_membership,
+            )
+        current_session_values = graph_session.scheduler_session.py_session.session_values
+        env = current_session_values[CompleteEnvironment]
+        return self._setup_bsp_connection(
+            union_membership=union_membership, env=env, options=goal_options
+        )
+
+    def _setup_bsp_connection(
+        self,
+        union_membership: UnionMembership,
+        env: Mapping[str, str],
+        options: OptionValueContainer,
     ) -> ExitCode:
+        """Setup the BSP connection file."""
+
+        build_root = BuildRoot()
+        bsp_conn_path = build_root.pathlib_path / ".bsp" / "pants.json"
+        if bsp_conn_path.exists():
+            print(
+                f"ERROR: A BSP connection file already exists at path `{bsp_conn_path}`. "
+                "Please delete that file if you intend to re-setup BSP in this repository."
+            )
+            return PANTS_FAILED_EXIT_CODE
+
+        bsp_dir = build_root.pathlib_path / ".pants.d" / "bsp"
+
+        bsp_scripts_dir = bsp_dir / "scripts"
+        bsp_scripts_dir.mkdir(exist_ok=True, parents=True)
+
+        bsp_logs_dir = bsp_dir / "logs"
+        bsp_logs_dir.mkdir(exist_ok=True, parents=True)
+
+        run_script_env_lines: list[str] = []
+        for env_var in options.runner_env_vars:
+            if "=" in env_var:
+                run_script_env_lines.append(env_var)
+            else:
```

Comment:
We've used this env pattern a few times (optional key-value, otherwise use current value), and it would be good to reuse some of it. Unfortunately, I can't seem to find any of the other locations (cc @kaos).

Reply:
The logic in [`CompleteEnvironment.get_subset`](https://github.com/pantsbuild/pants/blob/9c417d7a0cc545e023bb4b2b1b8a3395b2b4528b/src/python/pants/engine/environment.py#L37-L76) seems to be relevant.

Reply:
Can treat as a follow-on. I'd rather not refactor environment handling in this PR. Will add a TODO as suggestion.

Reply:
@stuhood perhaps in this area.. ? https://github.com/pantsbuild/pants/blob/79140750fa921bcc1d51722d2c161cc5236911bd/src/python/pants/backend/docker/util_rules/docker_build_env.py#L66-L69